README.md
[![Build status](https://ci.appveyor.com/api/projects/status/djfbr3frv1nstkjo?svg=true)](https://ci.appveyor.com/project/fdipuma/nhapi)

# nHapi.Core
NHapi is a .NET port of the original Java project [HAPI](http://hl7api.sourceforge.net/). This project is a simplified fork of the original [NHapi](https://github.com/duaneedwards/nHapi) project developed by [Duane Edwards](https://github.com/duaneedwards), which removes some Framework dependencies in order to improve portability in .NET Standard and .NET Core projects.

NHapi allows Microsoft .NET developers to easily use an HL7 2.x object model. This object model allows for parsing and encoding HL7 2.x data to/from Pipe Delimited or XML formats. A very handy program for use in the health care industry.

This project is **NOT** affiliated with the HL7 organization. This software just conforms to the HL7 2.x specifications.

**Key Benefits**

- Easy object model  
- Microsoft .NET Standard 2.0 library that conforms to HL7 2.1, 2.2, 2.3, 2.3.1, 2.4, 2.5, 2.5.1, 2.6, 2.7, 2.7.1, 2.8 and 2.8.1 specifications  
- Can take a pipe delimited or XML formatted HL7 2.x message and build the C# object model for use in code  
- Can take the C# HL7 object model and produce pipe delimited or XML formatted HL7  
- FREE! (You can't beat that price) and open source  
- Fast  

## Requirements

NHapi.Core currently targets .NET Standard 2.0

## Getting Started

The easiest way to get started using nHapi is to use the [NuGet package 'nHapi.Core'](https://www.nuget.org/packages/nHapi.Core/):

Using the package manager console withing visual studio, simply run the following command:

```
PM > Install-Package nHapi.Core
```

## Differences from the original nHapi project

In order to simplify dependencies, this for does not implement some features originally available in nHapi:

- Model Generation and source generation have been completely stripped off. This also removes the need to take a dependency on `System.Data.Odbc`.
- The configuration is now handled from code, in order to remove the use of `System.Configuration.ConfigurationManager` (which uses unavailable APIs for .NET Core projects).

The last point means that, in order to add a custom package version, you now cannot rely on using `app.config`. You should configure packages at application startup, instead:

For instance:

```c#
using System;
public class Program
{
    public static void Main()
    {
		PackageManager.Instance.AddCustomVersion("NHapi.Model.V22_ZSegments", "2.2.CustomZ");
        // other application code
    }
}

```

## [Change Log](https://github.com/duaneedwards/nHapi/blob/master/CHANGELOG.md)

CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

## Upcoming Release
###
-

## [2.6.0.0] - 2019-07-02
### Fixes
- Update project to support netstandard2.0 #61 **(Submitted by haydenhall and milkshakeuk)**
- Removed dependencies to .NET Framework assemblies to improve portability

## [2.5.0.6] - 2016-09-13
### Fixes
- Resolve #11 by reverting:
    - "Fix pipe parser and message iterator in handing white space in version and unexpected segments"
    - "Add unknown segments to the message not child groups."
    - "Fix pipe parser segment rep, and invalid field reps."
    - "Fix rep code to actually read using the GetField(i, rep)"

## [2.5.0.5] - 2016-07-27
### Fixes
- If encountering a DTM typed OBX result in a version of HL7 that does not have the DTM type, redirect to the DT type

## [2.5.0.4] - 2016-07-15
### Fixes  (Submitted by dib0)
- Small bug in milliseconds parsing CommonTM.cs #48

## [2.5.0.3] - 2016-07-15
### Fixes  (Submitted by Eric-Cadwell)
- TSComponentOne GetAsDate does not handle offset from UTC/GMT #38
- Whitespace after version number causes an issue #39
- Unexpected Segments cause later parsing issues #40
- Message escaping isn't functioning properly on all UNC strings #44

## [2.5.0.2] - 2016-07-15
### Added
- All versions: Repeating structures now have Enumerators, Add, Remove and RemoveAt methods generated

## [2.5.0.1] - 2016-07-14
### Breaking Changes
- Many breaking changes for v2.2, v2.3 and v2.3.1.  These are all due to fields not correctly being repeating or single fields, incorrect data types etc.
- All versions now successfully able to be code generated once more.

## [2.4.0.20] - 2016-07-13
### Added
- Initial implementation for HL7 version 2.8.1

## [2.4.0.19] - 2016-07-13
### Added
- Initial implementation for HL7 version 2.8

## [2.4.0.18] - 2016-07-13
### Added
- Initial implementation for HL7 version 2.7.1

## [2.4.0.17] - 2016-07-12
### Added
- Initial implementation for HL7 version 2.7

## [2.4.0.16] - 2016-07-12
### Added
- Initial implementation for HL7 version 2.6

## [2.4.0.15] - 2016-06-10
### Fixed
- Restore original behaviour for Custom Segment handling as reported in issue #28 (Submitted by lniedzielski)

## [2.4.0.14] - 2016-06-09
### Fixed
- Fix nHapi v2.5.1 not mapping ADT events correctly due to lack of having an eventmapping.properties embedded in the project (Submitted by rajputs6)

## [2.4.0.13] - 2016-05-19
### Fixed
- Fix incorrect version for all dlls
- Fix NHapi.Model.V251.dll not having it's version property set correctly

## [2.4.0.12] - 2016-04-06
### Fixed
- Add xml comments to nuget package (Submitted by vongillern)

## [2.4.0.11] - 2015-04-23
### Fixed
- HL7 Version 2.5.1 messages would not parse due to having incorrect namespaces.  Affected segments "DT, ST and TM". (Reported by lharless)

## [2.4.0.10] - 2015-03-15
### Fixed
- HL7 Version 2.1 messages would not parse due to not being listed as an acceptable base version. (Reported by taiji123)

## [2.4.0.9] - 2015-03-03
### Fixed
- Catch unhandled InvalidOperationExceptions when populating a HL7Exception's ERR segment using HL7Exception.populate().

## [2.4.0.8] - 2015-03-02
### Fixed
- Resolve EventMaps not being loaded properly.
- Update FT's SizeRule to be up to 64k in length.
- Include field offset component when rethrowing caught DataTypeExceptions whilst parsing.

## [2.4.0.7] - 2015-02-21
### Fixed
- Abstract and Varies segments were not having their Description field values populated. (Reported bv TheStephenStanton)

## [2.4.0.6] - 2015-02-21
### Fixed
- ORU_R01 messages for v2.5.1 incorrectly did not support the repeating, optional segments OBX, Timing QTY and Specimen.
- Moved regular expression field validation into it's own StrictValidation opt in validation context.
- Correctly escape \X0A\ as \n (thanks to riaz.ahmad)
- Each field / token was being extracted twice (Reported by AFAde)

## [2.4.0.5] - 2015-02-19
### Fixed
- Nuget Package wasn't deploying the 2.51 dll due to new nuget command line packaging and deployment.


NHapi20/NHapi.Base/PackageManager.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using NHapi.Base.Parser;

namespace NHapi.Base
{
	public class PackageManager
	{
		private static readonly PackageManager _instance = new PackageManager();
		private List<Hl7Package> _packages = new List<Hl7Package>();

		#region Constructors

		static PackageManager()
		{
		}

		private PackageManager()
		{
			LoadBaseVersions();
		}

		#endregion

		#region Properties

		public static PackageManager Instance
		{
			get { return _instance; }
		}

		internal IList<Hl7Package> GetAllPackages()
		{
			return _packages;
		}

		#endregion

		#region Methods

		private void LoadBaseVersions()
		{
			string[] versions = new string[] { "2.1", "2.2", "2.3", "2.3.1", "2.4", "2.5", "2.5.1", "2.6", "2.7", "2.7.1", "2.8", "2.8.1" };
			foreach (string version in versions)
			{
				string packageName = GetVersionPackageName(version);
				_packages.Add(new Hl7Package(packageName, version));
			}
		}

		/// <summary>
		/// Adds a custom package and version
		/// </summary>
		/// <param name="packageName"></param>
		/// <param name="version"></param>
		/// <exception cref="ArgumentNullException"></exception>
		public void AddCustomVersion(string packageName, string version)
		{
			if (packageName == null) throw new ArgumentNullException(nameof(packageName));
			if (version == null) throw new ArgumentNullException(nameof(version));

			if (_packages.Any(p => p.PackageName == packageName && p.Version == version))
				return;

			_packages.Add(new Hl7Package(packageName, version));

			// invalidate static cache of other components relying on PackageManager
			EventMapper.Instance.UseCache = false;
			DefaultModelClassFactory.UseCache = false;
		}

		internal bool IsValidVersion(string version)
		{
			version = version.ToUpper().Trim();
			foreach (Hl7Package package in _packages)
			{
				if (package.Version.ToUpper().Trim().Equals(version))
					return true;
			}
			return false;
		}

		#endregion

		#region Static GetVersion and GetVersionPackage

		/// <summary> Returns the path to the base package for model elements of the given version
		/// - e.g. "NHapi.Model.VXXX".
		/// This package should have the packages datatype, segment, group, and message
		/// under it. The path ends in with a slash.
		/// </summary>
		internal static String GetVersionPackagePath(String ver)
		{
			StringBuilder path = new StringBuilder("NHapi.Model.V");
			char[] versionChars = new char[ver.Length];
			SupportClass.GetCharsFromString(ver, 0, ver.Length, versionChars, 0);
			for (int i = 0; i < versionChars.Length; i++)
			{
				if (versionChars[i] != '.')
					path.Append(versionChars[i]);
			}
			path.Append("/");
			return path.ToString();
		}


		/// <summary> Returns the package name for model elements of the given version - e.g.
		/// "NHapi.Base.Model.v24.".  This method
		/// is identical to <code>getVersionPackagePath(...)</code> except that path
		/// separators are replaced with dots.
		/// </summary>
		internal static String GetVersionPackageName(String ver)
		{
			String path = GetVersionPackagePath(ver);
			String packg = path.Replace('/', '.');
			packg = packg.Replace('\\', '.');
			return packg;
		}

		#endregion
	}
}